### DISCUSSION TITLE
FRONTEND_RUNNER_GPU_LOCK implementation

### DISCUSSION THREAD

**Discussion by ale94mleon** [ale94mleon/BindFlow#16] [Ideas]
Although this help for debugging propose, it is not expose to user customization, the use of 1 GPU is hard-code in src/bindflow/orchestration/generate_scheduler.py (8a5dd8f). And for debugging propose the correct use of jobs and threads may be sufficient. 

As was discussed in #10, the real applications in the cluster should not be affected by this flag.

For now we can just let it like this, but keep this issue open for the future.

**Comment by LAndersen1** [comment]
The nice thing about this fix is that it does not limit the usage to 1 GPU (instead it uses all available GPUs, but only allows the execution of 1 job with the FRONTEND_RUNNER_GPU_LOCK flag). Also it only affects the execution if the resource FRONTEND_RUNNER_GPU_LOCK is specified via the command: `snakemake -resource FRONTEND_RUNNER_GPU_LOCK=1 ...`
An improvement would be to assign each job to a specific set of gpu_ids and thereby improve parallelism on systems with very many GPUs available.

**Comment by ale94mleon** [comment]
Thanks for bringing this up. Let's unpack it a bit further.

 Let's think in the following example. We have a workstation (or frontend) with 4 GPUs and, therefore, we can run, in principles, 4 of those rules that have the `FRONTEND_RUNNER_GPU_LOCK` (that in the end are the expensive ones). Because we build the snakemake command in `bindflow.orchestration.generate_scheduler.FrontEnd.build_snakemake` and in this method we say `--resources FRONTEND_RUNNER_GPU_LOCK=1`. We are saying that we have only one FRONTEND_RUNNER_GPU_LOCK resources when in the reality we have 4. And therefore instead of 4 expensive rules running in parallel, we will only have 1.

But here's the thing: this limitation isn't a big deal for us right now. BindFlow's main gig revolves around high-throughput tasks, mainly on clusters managed by SLURM, PBS, etc. In those setups, the cluster scheduler handles resource allocation like a pro, so we're not sweating over this restriction.

Your idea about dynamically assigning gpu_ids in Snakemake sounds cool, but it might be a bit overkill for us at the moment. 

 However, it's good to keep the discussion open for future possibilities we haven't thought of yet.
